alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -0,0 +1,457 @@
+  -- Part 2
+  (all (not . nonNative) utxoFees) ?! ScriptsNotPaidUTxO (UTxO (Map.filter nonNative utxoFees))
+  -- Part 3
+  (minfee pp tx <= theFee) ?! FeeNotBalancedUTxO (minfee pp tx) theFee
+  pure ()
+
+restrict :: Set (TxIn (Crypto era)) -> UTxO era -> UTxO era
+restrict txins (UTxO m) = UTxO (eval (txins ◁ m))
+
+-- ================================================================
+
+-- | The UTxO transition rule for the Alonzo eras.
+utxoTransition ::
+  forall era.
+  ( ValidateScript era,
+    Embed (Core.EraRule "PPUP" era) (AlonzoUTXO era),
+    Environment (Core.EraRule "PPUP" era) ~ PPUPEnv era,
+    State (Core.EraRule "PPUP" era) ~ PPUPState era,
+    Signal (Core.EraRule "PPUP" era) ~ StrictMaybe (Update era),
+    -- We leave Core.PParams abstract
+    UsesPParams era,
+    HasField "_minfeeA" (Core.PParams era) Natural,
+    HasField "_minfeeB" (Core.PParams era) Natural,
+    HasField "vldt" (Alonzo.TxBody era) ValidityInterval,
+    HasField "_keyDeposit" (Core.PParams era) Coin,
+    HasField "_poolDeposit" (Core.PParams era) Coin,
+    HasField "_minUTxOValue" (Core.PParams era) Coin,
+    HasField "_maxTxSize" (Core.PParams era) Natural,
+    HasField "_prices" (Core.PParams era) Prices,
+    HasField "_maxTxExUnits" (Core.PParams era) ExUnits,
+    -- We fix Core.Tx, Core.Value, Core.TxBody, and Core.TxOut
+    Core.TxOut era ~ Alonzo.TxOut era,
+    Core.Value era ~ Alonzo.Value (Crypto era),
+    Core.TxBody era ~ Alonzo.TxBody era,
+    Core.TxOut era ~ Alonzo.TxOut era
+  ) =>
+  TransitionRule (AlonzoUTXO era)
+utxoTransition = do
+  TRC (Shelley.UtxoEnv slot pp stakepools genDelegs, u, tx) <- judgmentContext
+  let Shelley.UTxOState utxo _deposits _fees ppup = u
+
+  let txb = txbody tx
+
+  inInterval slot (getField @"vldt" txb)
+    ?! OutsideValidityIntervalUTxO (getField @"vldt" txb) slot
+
+  txins @era txb /= Set.empty ?! InputSetEmptyUTxO
+
+  -- feesOK pp tx utxo
+
+  let minimumFee = minfee pp tx
+      txFee = getField @"txfee" txb
+  minimumFee <= txFee ?! FeeTooSmallUTxO minimumFee txFee
+
+  eval (txins @era txb ⊆ dom utxo)
+    ?! BadInputsUTxO (eval ((txins @era txb) ➖ (dom utxo)))
+
+  ni <- liftSTS $ asks networkId
+  let addrsWrongNetwork =
+        filter
+          (\a -> getNetwork a /= ni)
+          (fmap (getField @"address") $ toList $ getField @"outputs" txb)
+  null addrsWrongNetwork ?! WrongNetwork ni (Set.fromList addrsWrongNetwork)
+  let wdrlsWrongNetwork =
+        filter
+          (\a -> getRwdNetwork a /= ni)
+          (Map.keys . unWdrl . getField @"wdrls" $ txb)
+  null wdrlsWrongNetwork
+    ?! WrongNetworkWithdrawal
+      ni
+      (Set.fromList wdrlsWrongNetwork)
+
+  let consumed_ = consumed pp utxo txb
+      produced_ = Shelley.produced @era pp stakepools txb
+  consumed_ == produced_ ?! ValueNotConservedUTxO consumed_ produced_
+
+  -- process Protocol Parameter Update Proposals -- NOT SURE WHAT IS GOING On HERE
+  _ppup' <-
```

Comment:
This seems wrong, seeing as this is also called in the `UTXOS` rule

Reply:
As per fig. 11 of https://hydra.iohk.io/build/5463482/download/1/alonzo-changes.pdf there is no call to PPUP here

Reply:
I have removed it.